core/optaplanner-core-impl/src/main/java/org/optaplanner/core/config/solver/EnvironmentMode.java

```diff
@@ -81,62 +81,34 @@ public enum EnvironmentMode {
      * <p>
      * In multithreaded scenarios, this mode allows the use of work stealing and other non deterministic speed tricks.
      */
-    NON_REPRODUCIBLE;
+    NON_REPRODUCIBLE(false);
+
+    private final boolean asserted;
+
+    EnvironmentMode(boolean asserted) {
+        this.asserted = asserted;
+    }
 
     public boolean isAsserted() {
-        switch (this) {
-            case FULL_ASSERT:
-            case NON_INTRUSIVE_FULL_ASSERT:
-            case FAST_ASSERT:
-                return true;
-            case REPRODUCIBLE:
-            case NON_REPRODUCIBLE:
-                return false;
-            default:
-                throw new IllegalStateException("The environmentMode (" + this + ") is not implemented.");
-        }
+        return asserted;
     }
```

Comment:
This is another potentially interesting piece.
Since `isAsserted()` is now called on the hot path in groupBy, I need to make sure it is as fast as possible.